### PR TITLE
Preemptive priority in C++

### LINKED ISSUE #1947: priority scheduling (preemptive)  in scheduling algorithms C/C++.
I would like to work to on these algorithms in C/C++.
please assign me this issue. :)

### PR DESCRIPTION
**Pull request by pratu098**
Fixes #1947

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- 
- 

#### Languages Used:
- C++
- 

#### Files Added:
- Preemptive.cpp
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!
@asha15

### CODE REVIEW

File: Preemptive.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
+About:In priority scheduling, a number is assigned to each process that indicates its priority level. Lower the number, higher is the priority. In this type of scheduling algorithm, if a newer process arrives, that is having a higher priority than the currently running process, then the currently running process is preempted.
```
Comment (asha15) [inline Preemptive.cpp:1]: please break this into two sentences.

File: Preemptive.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    int intime;
+};
+
+// It is used to include all the valid and eligible
+// processes in the heap for execution. heapsize defines
+// the number of processes in execution depending on
+// the current time currentTime keeps a record of
+// the current CPU time.
+void insert(Process Heap[], Process value, int* heapsize,
+            int* currentTime)
```
Comment (asha15) [inline Preemptive.cpp:23]: take this to the previous line

File: Preemptive.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+{
+    int start = *heapsize, i;
+    Heap[*heapsize] = value;
+    if (Heap[*heapsize].intime == -1)
+        Heap[*heapsize].intime = *currentTime;
+    ++(*heapsize);
+
+    // Ordering the Heap
+    while (start != 0 && Heap[(start - 1) / 2].priority >
+                                  Heap[start].priority) {
```
Comment (asha15) [inline Preemptive.cpp:33]: take the `Heap[start].priorit)` into the previous line

File: Preemptive.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+void order(Process Heap[], int* heapsize, int start)
+{
+    int smallest = start;
+    int left = 2 * start + 1;
+    int right = 2 * start + 2;
+    if (left < *heapsize && Heap[left].priority <
+                            Heap[smallest].priority)
+        smallest = left;
+     if (right < *heapsize && Heap[right].priority <
+                           Heap[smallest].priority)
```
Comment (asha15) [inline Preemptive.cpp:53]: please state the if condition in one line

File: Preemptive.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        Heap[start] = temp;
+        order(Heap, heapsize, smallest);
+    }
+}
+
+// This function is used to find the process with
+// highest priority from the heap. It also reorders
+// the heap after extracting the highest priority process.
+Process extractminimum(Process Heap[], int* heapsize,
+                       int* currentTime)
```
Comment (asha15) [inline Preemptive.cpp:69]: take this to the previous line

File: Preemptive.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    return min;
+}
+
+// Compares two intervals according to staring times.
+int compare(Process p1, Process p2)
+{
+     if (p1.arrivalTime < p2.arrivalTime)
+        return 1;
+     else
+        return 0;
```
Comment (asha15) [inline Preemptive.cpp:88]: correct the indentation

**Comment by pratu098** [comment]
@asha15 I have done the required changes.

File: Preemptive.cpp [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    return min;
+}
+
+// Compares two intervals according to staring times.
+int compare(Process p1, Process p2)
+{
+     if (p1.arrivalTime < p2.arrivalTime)
+         return 1;
+     else
+         return 0;
```
Comment (asha15) [inline Preemptive.cpp:86]: correct the indentation

File: Preemptive.cpp [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+void scheduling(Process Heap[], Process array[], int n,int* heapsize, int* currentTime)
+{
+    if (heapsize == 0)
+        return;
+
+    Process min = extractminimum(Heap, heapsize, currentTime);
+    min.outtime = *currentTime + 1;
+    --min.burstTime;
+    printf("process id = %d\tcurrent_time = %d\n",
+           min.processID, *currentTime);
```
Comment (asha15) [inline Preemptive.cpp:100]: take this to previous line

**Comment by pratu098** [comment]
@asha15 Done the required changes.

**Pull request merged**